language/stdlib/modules/libra_system.move

```diff
@@ -70,94 +100,255 @@ module LibraSystem {
+        _  = Vector::swap_remove(&mut validator_set.validators, to_remove_index);
+
+        // Remove corresponding DiscoveryInfo from the discovery set, the indices should match
+        let discovery_set_ref = &mut borrow_global_mut<DiscoverySet>(0xD15C0).discovery_set;
+        _  = Vector::swap_remove(discovery_set_ref, to_remove_index);
+
+        set_validator_set(validator_set);
+        emit_discovery_set_change();
     }
 
-    // DiscoveryInfo public accessors
+    // Copies validator's config to ValidatorSet
+    // Callable by the LibraAssociation address only.
+    // Aborts if there are no changes to the config.
+    public fun update_validator_info(addr: address) acquires DiscoverySet {
+        Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+        let validator_set = get_validator_set();
 
-    public fun get_discovery_address(d: &DiscoveryInfo): &address {
-        &d.addr
+        let validator_index_vec = get_validator_index_(&validator_set.validators, addr);
+        Transaction::assert(!Vector::is_empty(&validator_index_vec), 19);
+
+        let validator_index = *Vector::borrow(&validator_index_vec, 0);
+        if (update_ith_validator_info_(&mut validator_set.validators, validator_index)) {
+            set_validator_set(validator_set);
+        };
+
+        if (update_ith_discovery_info_(&mut borrow_global_mut<DiscoverySet>(0xD15C0).discovery_set, validator_index)) {
+            emit_discovery_set_change();
+        };
     }
 
-    public fun get_validator_network_identity_pubkey(d: &DiscoveryInfo): &vector<u8> {
-        &d.validator_network_identity_pubkey
+    public fun force_update_validator_info_of_sender() acquires DiscoverySet {
+        force_update_validator_info(Transaction::sender());
     }
 
-    public fun get_validator_network_address(d: &DiscoveryInfo): &vector<u8> {
-        &d.validator_network_address
+    // This function can be called by the validator's operator,
+    // which is either the validator itself or the validator's delegate account.
+    // Aborts if there are no changes to the config.
+    public fun force_update_validator_info(validator_address: address) acquires DiscoverySet {
+        // Check the sender
+        Transaction::assert(ValidatorConfig::get_validator_operator_account(validator_address) == Transaction::sender(), 1);
+
+        let validator_set = get_validator_set();
+        let validator_index_vec = get_validator_index_(&validator_set.validators, validator_address);
+        Transaction::assert(!Vector::is_empty(&validator_index_vec), 19);
+        let validator_index = *Vector::borrow(&validator_index_vec, 0);
+
+        // TODO(valerini): add rate-limits for forced key rotations
+
+        // Copy validator config from validator's account and assert that there are changes to the config
+        if (update_ith_validator_info_(&mut validator_set.validators, validator_index)) {
+            set_validator_set(validator_set);
+        };
```

Comment:
Are we allowing modifications to the validator_set without calling a reconfiguration?